### PR TITLE
Graphql bundle targeting (App Interface pr check early exit support)

### PR DESCRIPTION
**Pull request by dwelch0**
In order for git partition sync to be included within app-interface PR check process, early exit should be supported. [Context on how app-interface sets previous bundle value](https://gitlab.cee.redhat.com/service/app-interface/-/blob/master/hack/manual_reconcile.sh#L31) 

These changes allow for app to query previous and current data bundles, compare the bundles for relevant attribute (`codeComponents.GitlabSync`) and `saas-git-partition-sync-producer` changes, and exit early if no changes are found. This check is triggered if optional env var `PREVIOUS_BUNDLE_SHA` is set

ticket: https://issues.redhat.com/browse/APPSRE-6544

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -40,7 +42,12 @@ func main() {
 		log.Fatalln(err)
 	}
 
-	// retrieve raw from graphql
+	// processed separately from env map above because optional w/ no default val
+	masterBundleSha := os.Getenv("MASTER_BUNDLE_SHA256")
```
Comment (geoberle) [inline main.go:46]: You don't have to adhere to the env var naming of our a-i PR check pipeline. Pick what makes sense, e.g. also a cli flag over an env var if that suits the UX of this tool better. 

You can always adapt in your hack runner func.

**Review by geoberle — COMMENTED** [review]
How are you controlling if you want to avoid early exit, e.g. in situations where you are promoting this integration as part of a PR.

**Comment by dwelch0** [comment]
@geoberle ah right, you raised this earlier but it just clicked what the implications are.

At the moment, the approach that seems most feasible would be to update the query file to look for this specific saas file and check for a change?

**Comment by geoberle** [comment]
> @geoberle ah right, you raised this earlier but it just clicked what the implications are.
> 
> At the moment, the approach that seems most feasible to me would be to update the query file to query saas files and then look for this specific saas file and check for a change?

many options are available to you. for q-r based integration the approach is the following

* see the print_cmd func in select-integrations.py - it has a way to figure out if early_exit is applicable in the context a PR
* if early_exit is applicable, it adds EARLY_EXIT=true to the CMD that is generated
* runners.sh uses that information to provide an env var with the the comparison bundle SHA
* an integration reacts to the presence of that env var (EARLY_EXIT_COMPARE_SHA) to run an early-exit check or not

you don't have to do it like this if it does not fit your integration. but if you do, you will get some things for free from the a-i hack scripts

**Comment by dwelch0** [comment]
@geoberle quick update on where things stand. I have refactored the PR check logic to utilize a different graphql query file. 

The new query file also pulls in saas files/resourceTemplates. With this data, the pr check not only looks for changes on the relevant `codeComponents.gitlabSyncs` but also looks for the specific saas file (uses env var to specify name of saas file to look for) and check that the targets are identical between bundles.

**Comment by dwelch0** [comment]
^ clean up after rebasing metrics PR

File: main.go [hunk 2]
```diff
@@ -120,3 +120,34 @@ func getEnvVars(vars map[string]string) (map[string]string, error) {
...
+			envVars["GRAPHQL_USERNAME"],
+			envVars["GRAPHQL_PASSWORD"],
+			prevBundleSha,
+			saasName,
+		)
+		if err != nil {
+			log.Fatalln(err)
+		} else if canExit {
+			log.Println("Relevant config is the same. Exiting early")
+			os.Exit(0)
```
Comment (geoberle) [inline main.go:150]: nit: this function would be easier to unit-test if you don't exit(0) here but instead return the decision about early-exit or not and let the main function exit conditionally.

File: main.go [hunk 3]
```diff
@@ -120,3 +120,34 @@ func getEnvVars(vars map[string]string) (map[string]string, error) {
 	}
 	return result, nil
 }
+
+func prCheckEarlyExit(envVars map[string]string) {
+	// indicates PR check when set
+	prevBundleSha := os.Getenv("PREVIOUS_BUNDLE_SHA")
```
Comment (geoberle) [inline main.go:126]: from this line i assume your strategy for the runners.sh in a-i is that you pass this env var if the MR supports it, so if only data or resource files change?

**Review by geoberle — APPROVED** [review]
i'm not familiar with this integration so i apologise that i can't review for correctness in detail.
the implementation for early exit looks legit. only minor nits you are free to ignore.
thanks for the hard work. LGTM

File: main.go (hunk 2 above)
Comment (dwelch0) [reply main.go:150]: Very good point. Will refactor

File: main.go (hunk 3 above)
Comment (dwelch0) [reply main.go:126]: Correct. Which I believe is okay for this functionality? 
The only attributes being queried are `codeComponent.gitlabSync` and `saasFile.resourceTemplate`. Both of which to my understanding solely exist within `data/`.

File: main.go (hunk 2 above)
Comment (dwelch0) [reply main.go:150]: After reviewing, going to revisit this in future PR. There's a larger overhaul to main that I'd like to complete that would work with these type of conditional checks better.

File: main.go (hunk 3 above)
Comment (geoberle) [reply main.go:126]: thx

**Pull request merged**